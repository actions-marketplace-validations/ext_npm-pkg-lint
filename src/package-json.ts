import PackageJson from "./types/package-json";
import { Message } from "./message";
import { Result } from "./result";
import { nonempty, present, typeArray, typeString, validUrl } from "./validators";

/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
type validator = (key: string, value: any) => void;

const fields: Record<string, validator[]> = {
	description: [present, typeString, nonempty],
	keywords: [present, typeArray, nonempty],
	homepage: [present, typeString, validUrl],
	bugs: [present, validUrl],
	license: [present, typeString, nonempty],
	author: [present, nonempty],
	repository: [present, validUrl],
};

export async function verifyPackageJson(pkg: PackageJson, filePath: string): Promise<Result> {
	const messages: Message[] = [];

	for (const [field, validators] of Object.entries(fields)) {
		try {
			for (const validator of validators) {
				validator(field, pkg[field]);
			}
		} catch (error) {
			messages.push({
				ruleId: "package-json-fields",
				severity: 2,
				message: error.message,
				line: 1,
				column: 1,
			});
		}
	}

	return {
		messages,
		filePath,
		errorCount: messages.length,
		warningCount: 0,
	};
}
