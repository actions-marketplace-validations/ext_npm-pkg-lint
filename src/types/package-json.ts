interface UrlObject {
	type?: string;
	name?: string;
	url?: string;
	email?: string;
}

export default interface PackageJson {
	name: string;
	version: string;
	description?: string;
	keywords?: string[];
	homepage?: string;
	bugs?: string | UrlObject;
	license?: string;
	author?: string | UrlObject;
	repository?: string | UrlObject;
	main?: string;
	browser?: string | Record<string, string | false>;
	module?: string;
	"jsnext:main"?: string;
	typings?: string;
	bin?: string | Record<string, string>;
	man?: string | string[];
	dependencies?: Record<string, string>;
	devDependencies?: Record<string, string>;
	peerDependencies?: Record<string, string>;
	bundledDependencies?: Record<string, string>;
	optionalDependencies?: Record<string, string>;

	[key: string]:
		| string
		| string[]
		| Record<string, string>
		| Record<string, string | false>
		| UrlObject;
}
