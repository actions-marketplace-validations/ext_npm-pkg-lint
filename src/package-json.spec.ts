import { verifyPackageJson } from "./package-json";
import PackageJson from "./types/package-json";

let pkg: PackageJson;

beforeEach(() => {
	pkg = {
		name: "mock-pkg",
		version: "1.2.3",
		description: "description",
		keywords: ["test"],
		homepage: "https://example.net",
		bugs: "https://example.net",
		license: "UNLICENSED",
		author: "Fred Flintstone <fred.flintstone@example.net>",
		repository: "https://git.example.net/test-case.git",
	};
});

it("should not return errors if package.json is well formed (strings only)", async () => {
	expect.assertions(1);
	const result = await verifyPackageJson(pkg, "package.json");
	expect(result.messages).toEqual([]);
});

describe("description", () => {
	it("should return error if not set", async () => {
		expect.assertions(1);
		delete pkg.description;
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});

	it("should return error if not string", async () => {
		expect.assertions(1);
		pkg.description = (12 as unknown) as string;
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});

	it("should return error if empty", async () => {
		expect.assertions(1);
		pkg.description = "";
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});
});

describe("keywords", () => {
	it("should return error if not set", async () => {
		expect.assertions(1);
		delete pkg.keywords;
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});

	it("should return error if not string", async () => {
		expect.assertions(1);
		pkg.keywords = (12 as unknown) as string[];
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});

	it("should return error if empty", async () => {
		expect.assertions(1);
		pkg.keywords = [];
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});

	it("should return error if items are empty", async () => {
		expect.assertions(1);
		pkg.keywords = [""];
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});
});

describe("homepage", () => {
	it("should return error if not set", async () => {
		expect.assertions(1);
		delete pkg.homepage;
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});

	it("should return error if not string", async () => {
		expect.assertions(1);
		pkg.homepage = (12 as unknown) as string;
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});

	it("should return error if not valid url", async () => {
		expect.assertions(1);
		pkg.homepage = "foobar";
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});
});

describe("bugs", () => {
	it("should return error if not set", async () => {
		expect.assertions(1);
		delete pkg.bugs;
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});

	it("should return error if not valid url", async () => {
		expect.assertions(1);
		pkg.bugs = "foobar";
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});
});

describe("license", () => {
	it("should return error if not set", async () => {
		expect.assertions(1);
		delete pkg.license;
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});

	it("should return error if not string", async () => {
		expect.assertions(1);
		pkg.license = (12 as unknown) as string;
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});

	it("should return error if empty", async () => {
		expect.assertions(1);
		pkg.license = "";
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});
});

describe("author", () => {
	it("should return error if not set", async () => {
		expect.assertions(1);
		delete pkg.author;
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});

	it("should return error if empty string", async () => {
		expect.assertions(1);
		pkg.author = "";
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});

	it("should return error if empty object", async () => {
		expect.assertions(1);
		pkg.author = {};
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});
});

describe("repository", () => {
	it("should return error if not set", async () => {
		expect.assertions(1);
		delete pkg.repository;
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});

	it("should return error if not valid url", async () => {
		expect.assertions(1);
		pkg.repository = "foobar";
		const result = await verifyPackageJson(pkg, "package.json");
		expect(result.messages).toMatchSnapshot();
	});
});
