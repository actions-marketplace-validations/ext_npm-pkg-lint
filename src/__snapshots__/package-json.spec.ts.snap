// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`author should return error if empty object 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"author\\" must not be empty",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`author should return error if empty string 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"author\\" must not be empty",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`author should return error if not set 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"author\\" must be set",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`bugs should return error if not set 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"bugs\\" must be set",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`bugs should return error if not valid url 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"bugs\\" must be a valid url (https only)",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`description should return error if empty 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"description\\" must not be empty",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`description should return error if not set 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"description\\" must be set",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`description should return error if not string 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"description\\" must be string",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`homepage should return error if not set 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"homepage\\" must be set",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`homepage should return error if not string 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"homepage\\" must be string",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`homepage should return error if not valid url 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"homepage\\" must be a valid url (https only)",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`keywords should return error if empty 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"keywords\\" must not be empty",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`keywords should return error if items are empty 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"keywords[0]\\" must not be empty",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`keywords should return error if not set 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"keywords\\" must be set",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`keywords should return error if not string 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"keywords\\" must be array",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`license should return error if empty 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"license\\" must not be empty",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`license should return error if not set 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"license\\" must be set",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`license should return error if not string 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"license\\" must be string",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`repository should return error if not set 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"repository\\" must be set",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`repository should return error if not valid url 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "\\"repository\\" must be a valid url (https only)",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;
